### PR TITLE
TSCH: speed up joining when a custom join hopping sequence is defined

### PR DESCRIPTION
**Pull request by atiselsts**
**The problem:** TSCH joining speed is inversely proportional to the number of channels to scan. If a large number of channels is used (for example 64, which might be needed due to regulatory limitations in sub-GHZ operation) then joining can be very slow. Configuring a join hopping sequence with a smaller number of channels does not help as EB packets are still sent out using the full channel hopping sequence.

**The solution:** if a custom joining sequence is defined, only send out EB packets using channels in that joining sequence.
This only affects networks with custom configuration: the default behavior is not changed.

When comparing the simple-node example with custom configuration (number of channels = 16, numbrer of join channels = 2) there is a marked improvement in speed. After the patch is takes 33 seconds for all nodes to join the network, before the patch: 300 seconds.

This is based on work done with Cysca Technologies.

### CODE REVIEW

**Comment by atiselsts** [comment]
I acknowledge this PR may make things worse in terms of EB collisions. However, it does not change the default behavior, and if someone changes the config to enable this feature, I think its fair to assume that they know what they are doing and can understand the issue. There are ways to work around the EB collision issue, you know, for example by making the EB more frequent. 

I totally agree on using multiple channel offsets for sending EBs, but it's not really directly relevant to this PR. Anyway, I tried doing this in once, and did not see much performance improvement for that particular setup, so it was sort of forgotten. However in some networks it might make a big difference and I don't think there are any networks. In fact let me remake #1673 with this in mind.

The EB in RPL messages is a feature that changes the standard behavior a lot and probably also complicates the code a lot, so I don't think it will ever be included in Contiki-NG. Worth investigating for sure if your goal is not to be standard compatible.

**Review by arurke — CHANGES_REQUESTED** [review]
I tested this patch some time ago and saw the same great benefit to join-time.

**Comment by arurke** [comment]
@atiselsts Should we aim for v4.8 here? I am happy to merge after the comment has been addressed.

File: os/net/mac/tsch/tsch-slot-operation.c [hunk 1]
```diff
@@ -335,6 +335,31 @@ tsch_schedule_slot_operation(struct rtimer *tm, rtimer_clock_t ref_time, rtimer_
     RTIMER_BUSYWAIT_UNTIL_ABS(0, ref_time, offset); \
   } while(0);
 /*---------------------------------------------------------------------------*/
+/*
+ * Check whether the current channel is in the join hopping sequence.
+ * If a custom join hopping sequence is defined, EB packets are only sent using that sequence.
```
[the remark this answers is not in the document]
Comment (atiselsts) [reply os/net/mac/tsch/tsch-slot-operation.c:340]: Right, I will change this in the few next days.

**Comment by atiselsts** [comment]
Updated now.

**Review by arurke — APPROVED** [review]

**Pull request merged**